src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs

```diff
@@ -0,0 +1,1378 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace Microsoft.CodeAnalysis.Semantics
+{
+    internal static partial class CSharpOperationFactory
+    {
+        private static readonly ConditionalWeakTable<BoundNode, IOperation> s_cache = new ConditionalWeakTable<BoundNode, IOperation>();
+
+        public static IOperation Create(BoundNode boundNode)
+        {
+            if (boundNode == null)
+            {
+                return null;
+            }
+
+            return s_cache.GetValue(boundNode, n => CreateInternal(n));
+        }
+
+        private static IOperation CreateInternal(BoundNode boundNode)
+        {
+            switch (boundNode)
+            {
+                case BoundDeconstructValuePlaceholder boundDeconstructValuePlaceholder:
```

Comment:
while this pattern matching is attractive because it doesn't require test+cast, it's also really expensive as every node may require dozens of type checks.

1. is there a visitor for bound nodes we can use to do this work?
2. if not, i think we should just use a regular Kind switch so that we don't do tons of type checks for every operation we create.


--

Note: this isn't a hypothetical concern.  I've run into i several time in throughput oriented domains.  type-switchs are great when small, or when you need to them rarely.  However, if we're converting tons of nodes over to operations, this isn't desirable.